src/Functions/parseDateTime.cpp

```diff
@@ -0,0 +1,1766 @@
+    };
+
+    const std::unordered_map<String, std::pair<String, Int32>> monthMap{
+        {"jan", {"uary", 1}},
+        {"feb", {"ruary", 2}},
+        {"mar", {"rch", 3}},
+        {"apr", {"il", 4}},
+        {"may", {"", 5}},
+        {"jun", {"e", 6}},
+        {"jul", {"y", 7}},
+        {"aug", {"ust", 8}},
+        {"sep", {"tember", 9}},
+        {"oct", {"ober", 10}},
+        {"nov", {"ember", 11}},
+        {"dec", {"ember", 12}},
+    };
+
+    /// key: month, value: total days of current month if current year is leap year.
+    constexpr Int32 leapDays[] = {0, 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+    /// key: month, value: total days of current month if current year is not leap year.
+    constexpr Int32 normalDays[] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
+
+    /// key: month, value: cumulative days from January to current month(inclusive) if current year is leap year.
+    constexpr Int32 cumulativeLeapDays[] = {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366};
+
+    /// key: month, value: cumulative days from January to current month(inclusive) if current year is not leap year.
+    constexpr Int32 cumulativeDays[] = {0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365};
+
+    /// key: year, value: cumulative days from epoch(1970-01-01) to the first day of current year(exclusive).
+    constexpr Int32 cumulativeYearDays[]
+        = {0,     365,   730,   1096,  1461,  1826,  2191,  2557,  2922,  3287,  3652,  4018,  4383,  4748,  5113,  5479,  5844,  6209,
+           6574,  6940,  7305,  7670,  8035,  8401,  8766,  9131,  9496,  9862,  10227, 10592, 10957, 11323, 11688, 12053, 12418, 12784,
+           13149, 13514, 13879, 14245, 14610, 14975, 15340, 15706, 16071, 16436, 16801, 17167, 17532, 17897, 18262, 18628, 18993, 19358,
+           19723, 20089, 20454, 20819, 21184, 21550, 21915, 22280, 22645, 23011, 23376, 23741, 24106, 24472, 24837, 25202, 25567, 25933,
+           26298, 26663, 27028, 27394, 27759, 28124, 28489, 28855, 29220, 29585, 29950, 30316, 30681, 31046, 31411, 31777, 32142, 32507,
+           32872, 33238, 33603, 33968, 34333, 34699, 35064, 35429, 35794, 36160, 36525, 36890, 37255, 37621, 37986, 38351, 38716, 39082,
+           39447, 39812, 40177, 40543, 40908, 41273, 41638, 42004, 42369, 42734, 43099, 43465, 43830, 44195, 44560, 44926, 45291, 45656,
+           46021, 46387, 46752, 47117, 47482, 47847, 48212, 48577, 48942, 49308, 49673};
+
+    struct DateTime
+    {
+        /// If both week_date_format and week_date_format is false, date is composed of year, month and day
+        Int32 year = 1970; /// year, range [1970, 2106]
+        Int32 month = 1; /// month of year, range [1, 12]
+        Int32 day = 1; /// day of month, range [1, 31]
+
+        Int32 week = 1; /// ISO week of year, range [1, 53]
+        Int32 day_of_week = 1; /// day of week, range [1, 7], 1 represents Monday, 2 represents Tuesday...
+        bool week_date_format
```

Comment:
As far as I see, `week_date_format` and `day_of_year_format` can take combinations
- week_date_format = true, day_of_year_format = false, or
- week_date_format = false, day_of_year_format = true, or
- week_date_format = false, day_of_year_format = false

I think the logic would benefit from an enum class with three states instead.

Reply:
It makes sense. DateTime will be shorter in memory, maybe there are some performance benefits.